public/client/TracyCallstack.cpp

```diff
@@ -90,6 +90,124 @@ extern "C" const char* ___tracy_demangle( const char* mangled )
+        if( !entry )
+        {
+            //printf("* addr not found: %p (%d entries) refreshing (m_numberOfRefreshes: %d)\n",
+            //       address, m_images->size(), m_numberOfRefreshes);
+            Refresh();
+            return GetImageEntryForAddress( address );
+        }
+        return entry;
+    }
+
+    typedef void (*RefreshCallback)();
+    void SetOnRefreshCallback( RefreshCallback callback )
+    {
+        m_onRefreshCallback = callback;
+    }
+
+private:
+    tracy::FastVector<ImageEntry>* m_images;
+    const char* m_imageName = nullptr;
+    RefreshCallback m_onRefreshCallback;
+    uint32_t m_numberOfRefreshes = 0;
+
+    static int Callback( struct dl_phdr_info* info, size_t size, void* data ) 
+    {
+        ImageCache* cache = reinterpret_cast<ImageCache*>( data );
+
+        ImageEntry* image = cache->m_images->push_next();
+        image->m_startAddress = reinterpret_cast<void*>( info->dlpi_addr );
+        const uint32_t headerCount = info->dlpi_phnum;
+        assert( headerCount > 0);
+        image->m_endAddress = reinterpret_cast<void*>( info->dlpi_addr + 
+            info->dlpi_phdr[info->dlpi_phnum - 1].p_vaddr + info->dlpi_phdr[info->dlpi_phnum - 1].p_memsz);
+
+        // the base executable name isn't provided when iterating with dl_iterate_phdr,
+        // so we must get it in an alternative way and cache it
+        if( info->dlpi_name && info->dlpi_name[0] != '\0' )
+        {
+            image->m_name = info->dlpi_name;
+        }
+        else
+        {
+            if( !cache->m_imageName )
+            {
+                Dl_info dlInfo;
+                if( dladdr( (void *)info->dlpi_addr, &dlInfo ) )
+                {
+                    cache->m_imageName = dlInfo.dli_fname;
+                }
+            }
+            image->m_name = cache->m_imageName;
+        }
+
+        //printf("\tdl_iterate_phdr:'%s', start:%p, end:%p (headerCount: %d)'\n",
+        //       image->m_name, image->m_startAddress, image->m_endAddress, headerCount);
+
+        return 0;
+    }
+
+    void Refresh()
+    {
+        m_images->clear();
+
+        dl_iterate_phdr( Callback, this );
+        
+        std::sort( m_images->begin(), m_images->end(), 
+            []( const ImageEntry& lhs, const ImageEntry& rhs ) { return lhs.m_startAddress > rhs.m_startAddress; } );
+
+        if( m_onRefreshCallback ) m_onRefreshCallback();
+
+        m_numberOfRefreshes++;
+    }
+
+    const ImageEntry* GetImageEntryForAddress( void* address ) const 
+    {
+        auto it = std::lower_bound( m_images->begin(), m_images->end(), address, 
+            []( const ImageEntry& lhs, const void* rhs ) { return lhs.m_startAddress > rhs; } );
+
+        if( it != m_images->end() && address < it->m_endAddress )
+        {
+            return &(*it);
+        }
+        return nullptr;
+    }
+};
+#endif //#ifdef TRACY_USE_IMAGE_CACHE
+
 namespace tracy
```

Comment:
Is there a reason for keeping the above code outside the `tracy` namespace?

Reply:
not really, moved it under